jfix-zookeeper/src/main/kotlin/ru/fix/zookeeper/instance/registry/ServiceInstanceIdRegistry.kt

```diff
@@ -0,0 +1,196 @@
+     * If this is client with host and port, that already contains in instance's znodes, and timeout of this lock expired,
+     * this situation seen as reconnect of instance, and previous instance id will be got.
+     * Generate instance id, this id should be in range 1..maxInstancesCount, assertion error will be thrown otherwise.
+     * If unsuccessful, then log error
+     */
+    private fun initInstanceId() {
+        for (i in 1..config.countRegistrationAttempts) {
+            val alreadyRegisteredInstanceIds = curatorFramework.children.forPath(config.serviceRegistrationPath)
+            val preparedInstanceId = resolvePreviousConnection(alreadyRegisteredInstanceIds)
+                    ?: instanceIdGenerator.nextId(alreadyRegisteredInstanceIds)
+
+            val instanceIdPath = "${config.serviceRegistrationPath}/$preparedInstanceId"
+            val instanceIdData = InstanceIdData(config.serviceName, Instant.now(), host = config.host, port = config.port)
+            val lockIdentity = LockIdentity(preparedInstanceId, instanceIdPath, Marshaller.marshall(instanceIdData))
+
+            val result = lockManager.tryAcquire(lockIdentity) { lock ->
+                logger.debug("Failed to initialize service id registry and generate instance id. " +
+                        "Lock id: ${Marshaller.marshall(lock)}. " +
+                        "Current registration node state: " +
+                        ZkTreePrinter(curatorFramework).print(config.serviceRegistrationPath, true)
+                )
+            }
+            if (result) {
+                this.instanceIdPath = instanceIdPath
+                this.instanceId = preparedInstanceId
+                this.lockIdentity = lockIdentity
+
+                logger.info("Instance started with id=$instanceId")
+                break
+            } else if (i == config.countRegistrationAttempts) {
+                logger.error("Failed to initialize instance id registry and get instance id. " +
+                        "Limit=${config.countRegistrationAttempts} of instance id registration reached. " +
+                        "Last try to get instance id was instanceId=$instanceId. " +
+                        "Current registration node state: " +
+                        ZkTreePrinter(curatorFramework).print(config.serviceRegistrationPath, true)
+                )
+            }
+        }
+    }
+
+    /**
+     * @return previous instance id if host and port of client same with data in zk node
+     * and lock of this instance id expired, and null otherwise
+     */
+    private fun resolvePreviousConnection(alreadyRegisteredInstanceIds: List<String>): String? {
+        alreadyRegisteredInstanceIds.forEach ids@{
+            val path = "${config.serviceRegistrationPath}/$it"
+            val lockData = Marshaller.unmarshall(
+                    curatorFramework.data.forPath(path).toString(Charsets.UTF_8),
+                    object : TypeReference<LockData>() {}
```

Comment:
You are breaking encapsulation. Why ServiceInstanceIdRegistry suddenly `knows` about internal representation of PersistentExpiringDistributedLock? It is bette that ServiceInstanceIdRegistry works only though ServiceInstanceIdRegistry api and do not couple with internals of PersistentExpiringDistributedLock.

Reply:
outdated